Flag tracker

Malkierian opened the pull request:
Switches check tracker over to `OnFlagSet` and `OnSceneFlagSet`, with some flag corrections in rcObjects. Still manually handles several vanilla "checks", like medallions, most songs, etc. This allows complete removal of delayed checking and check queues, making tracking instantaneous. Should now properly handle virtually all vanilla and MQ non-rando checks (though there are undoubtedly a few left that have bad values). Also changes tracker to always spoil all areas when not a randomizer save.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893582.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893583.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893584.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893585.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893586.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893587.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1078893588.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Hook into flagset hooks for processing check tracking. Has some manual workarounds (some breaks still need to be found).
Remove areaChecks and looping functionality as redundant.

Malkierian committed:
Additional vanilla handling.

Malkierian committed:
Fix tracker not showing MQ checks in MQ non-rando.
Fix tracker marking non-MQ variants of dungeon checks (e.g. map chest, etc) when collecting in MQ.

Malkierian committed:
Set all areas to spoiled if not rando.

Malkierian committed:
Merge branch 'develop-macready' of github.com:Malkierian/Shipwright into flag-tracker

Malkierian committed:
Revert attempt to spoil in randomizer based on MQ dungeon count as I don't know how that works at the moment.

Malkierian committed:
Restore and update spoiling based on MQ dungeon settings (none, selection, or count of 12).
Fix Anju As Adult check.

Malkierian committed:
Remove Anchor-specific code :baguette:

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -360,27 +341,9 @@ bool vector_contains_scene(std::vector<SceneID> vec, const int16_t scene) {
-        }
-        auto rco = areaChecks.at(checkCounter);
-        return EvaluateCheck(rco);
+void ClearAreaTotals() {
```
Looks like you left this one. This same'ish block of code is in the teardown so I'd either remove this function or call this from teardown and add the `vec.clear();`

Malkierian replied:
areaChecksGotten is the check counter for areas judging completion. Has nothing to do with the previous checkAreas functionality.

garrettjoecox replied:
I get that, but this function isn't used anywhere, should it be?

Malkierian replied:
Oh, I see. I never got around to replacing the code in `Teardown()` with it.

garrettjoecox replied:
This function is missing the `vec.clear();` that the teardown has, does that matter?

Malkierian replied:
LOL, yes, probably should.

Malkierian committed:
Use `ClearAreaTotals()` in `Teardown()` instead of the duplicate code there.

Malkierian committed:
Update to `ClearAreaChecksAndTotals()` with `vec.clear()` added.

Malkierian committed:
Fix type spoiling again. Now spoils on 0 MQ dungeons, not rando, if the option is enabled in check tracker settings, selection, or set number of 12.
Fix vanilla checks being marked collected in MQ dungeons.

Malkierian committed:
Fix 100 GS check.

Malkierian committed:
ACTUALLY fix 100 GS: change flag type to `RandomizerInf()` in `item_location.cpp`, add RC to RandoInf for it to the table. Also don't send GI for flag if father, falsely triggers ZR frogs minigame.

Malkierian committed:
Fix gravedigging tour tracking.

aMannus commented on soh/src/code/z_message_PAL.c:
```diff
@@ -3384,7 +3384,7 @@ void Message_Update(PlayState* play) {
             sLastPlayedSong = 0xFF;
             osSyncPrintf("OCARINA_MODE=%d   chk_ocarina_no=%d\n", play->msgCtx.ocarinaMode, msgCtx->unk_E3F2);
-            CheckTracker_OnMessageClose();
+            // TODO: OnMessageClose hook
```
What's the context here? Could we get it clarified in the comment what this is going to be for?

Malkierian replied:
I used to have a direct call to a function in the check tracker namespace for tracking message closing as I didn't want wait for a GI hook at the time. I don't kneed it anymore for check tracker, but that's where the GI hook should go in the future.

aMannus commented on soh/soh/Enhancements/randomizer/3drando/item_location.cpp:
```diff
@@ -24,11 +24,11 @@ void LocationTable_Init() {
     //Lost Woods
     locationTable[LW_NEAR_SHORTCUTS_GROTTO_CHEST]        = ItemLocation::Chest      (RC_LW_NEAR_SHORTCUTS_GROTTO_CHEST,             0x3E, 0x14, "LW Near Shortcuts Grotto Chest",       LW_NEAR_SHORTCUTS_GROTTO_CHEST,        BLUE_RUPEE,                {},                                                                                                    SpoilerCollectionCheckGroup::GROUP_LOST_WOODS) ...
-    locationTable[LW_SKULL_KID]                          = ItemLocation::Base       (RC_LW_SKULL_KID,                               0x5B,       "LW Skull Kid",                         LW_SKULL_KID,                          PIECE_OF_HEART,            {},                                                   SpoilerCollectionCheck::ItemGetInf(30),          SpoilerCollectionCheckGroup::GROUP_LOST_WOODS) ...
+    locationTable[LW_SKULL_KID]                          = ItemLocation::Base       (RC_LW_SKULL_KID,                               0x5B,       "LW Skull Kid",                         LW_SKULL_KID,                          PIECE_OF_HEART,            {},                                                   SpoilerCollectionCheck::ItemGetInf(22),          SpoilerCollectionCheckGroup::GROUP_LOST_WOODS) ...
```
Not really for this PR, but is this on the radar for Rando V3 to get enums for these?

Malkierian replied:
Yes, I believe the plan is for everything to have a randomizer inf at some point that is triggered in rando so we can track them separately from the in-game flags.

Malkierian replied:
I do want to clarify on this, though, it would be useful to still keep track of these, as they're useful for non-rando tracking, if the randomizer infs aren't going to be flipped both in and out of rando.

aMannus replied:
Wasn't the plan to move back to vanilla flags as much as possible? Anyway, yeah not really a conversation to hold here, just saw it and figured I'd mention it.

aMannus reviewed (approved):
Can't say I fully grasp the item tracker code, but looks ok to me? 2 small comments.

Archez commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,119 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
     }
-    if (scene == SCENE_DESERT_COLOSSUS && (gSaveContext.entranceIndex == 485 || gSaveContext.entranceIndex == 489)) {
-        checkAreas.push_back(RCAREA_SPIRIT_TEMPLE);
+    if (sceneNum == 83 && flag == 31 && flagType == FLAG_SCENE_COLLECTIBLE) { // Gravedigging tour special case
```
Use scene enum

Archez commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,119 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
+            break;
+        case FLAG_ITEM_GET_INF:
+            if (!IS_RANDO) {
+                if (flag == 30) {
```
I think we should use the flag enums where ever we can, even if they aren't named friendly

i.e. here would be `ITEMGETINF_OBTAINED_STICK_UPGRADE_FROM_STAGE`

Archez commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,119 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
-        checkAreas.push_back(checkArea);
-        checkCollected = true;
+    for (auto [rc, rcObj] : RandomizerCheckObjects::GetAllRCObjects()) {
+        if (!IS_RANDO && ((rcObj.vOrMQ == RCVORMQ_MQ && !IS_MASTER_QUEST) || (rcObj.vOrMQ == RCVORMQ_VANILLA && IS_MASTER_QUEST)) ||
+            IS_RANDO && (OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(rcObj.sceneId) && rcObj.vOrMQ == RCVORMQ_VANILLA) || 
+                         !OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(rcObj.sceneId) && rcObj.vOrMQ == RCVORMQ_MQ){
```
The readability of this could be improved if you wrapped the two `IS_RANDO` parts in parenthesis. Also I think there is a bug in the existing parens too maybe? (the last part is being `||` at the top level, instead of within the `IS_RANDO` part)

This is what I think it should be with proper parens and then clang-formatted.

```suggestion
        if ((!IS_RANDO && ((rcObj.vOrMQ == RCVORMQ_MQ && !IS_MASTER_QUEST) ||
                           (rcObj.vOrMQ == RCVORMQ_VANILLA && IS_MASTER_QUEST))) ||
            (IS_RANDO && ((OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(rcObj.sceneId) &&
                           rcObj.vOrMQ == RCVORMQ_VANILLA) ||
                          !OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(rcObj.sceneId) &&
                              rcObj.vOrMQ == RCVORMQ_MQ))) {
```

Malkierian replied:
These are all done now.

Malkierian committed:
Fix membership card check tracking.
Change scene and flag values to any existing enums.
Clarifying formatting for the checking loop vOrMQ conditions.

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -560,29 +514,12 @@ void CheckTrackerFrame() {
-            checkLoops = 0;
+    if (gSaveContext.checkTrackerData[RC_ZR_MAGIC_BEAN_SALESMAN].status != RCSHOW_COLLECTED &&
+        gSaveContext.checkTrackerData[RC_ZR_MAGIC_BEAN_SALESMAN].status != RCSHOW_SAVED) {
+        if (BEANS_BOUGHT >= 10) {
```
Doesn't need to happen now but this should probably use the onAmmoChange hook I made for something else

Malkierian replied:
Yeah, if that handles beans as well, that would certainly be better than a check every frame. I'll put a TODO in for when that gets added.

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -593,7 +530,7 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
     // Vanilla special item checks
     if (!IS_RANDO) {
```
I think you're already aware of this, but come rando v3 all of these should be handled when we adjust the flags in our rando table

Malkierian replied:
You're referring to the fact that we'll be consolidating flags, and whatever flag each check is registered with will always be flipped, regardless of rando or not?

garrettjoecox replied:
Right

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,124 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
+            checkMatchType = SpoilerCollectionCheckType::SPOILER_CHK_EVENT_CHK_INF;
+            break;
+        case FLAG_INF_TABLE:
+            if (flag == INFTABLE_190) {
+                SetCheckCollected(RC_GF_HBA_1000_POINTS);
+            } else if (flag == INFTABLE_11E) {
+                SetCheckCollected(RC_GC_ROLLING_GORON_AS_CHILD);
+            } else if (flag == INFTABLE_GORON_CITY_DOORS_UNLOCKED) {
+                SetCheckCollected(RC_GC_ROLLING_GORON_AS_ADULT);
+            } else if (flag == INFTABLE_139) {
+                SetCheckCollected(RC_ZD_KING_ZORA_THAWED);
+            } else if (flag == INFTABLE_191) {
+                SetCheckCollected(RC_MARKET_LOST_DOG);
+            }
```
Why don't we just rely on the ItemLocation flags for these? looks like they are all offset a bit but easy adjustments

Malkierian replied:
Because I don't know how to translate the `SpoilerCheckCollectionCheck::InfTable` type into the values I actually get from the hook. If you could enlighten me, I would be more than willing to put them in there.

garrettjoecox replied:
It's the reverse of the bit math you'll find in the Flags_Set functions, but since there's only 5 of these anyway, it's probably easiest to just update `SpoilerCheckCollectionCheck::InfTable` to take one argument that is the flag we use.

Malkierian replied:
Sounds like a v3 thing, though, right?

garrettjoecox replied:
Yeah thats fine

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,124 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
+                    SetCheckCollected(RC_LW_DEKU_SCRUB_NEAR_BRIDGE);
+                } else if (flag == INFTABLE_193) {
+                    SetCheckCollected(RC_LW_DEKU_SCRUB_GROTTO_FRONT);
+                }
```
missing returns, doesn't look like it matters but consistency

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,124 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
+        }
+        SpoilerCollectionCheck scCheck = Location(rc)->GetCollectionCheck();
+        SpoilerCollectionCheckType scCheckType = scCheck.type;
+        if (checkMatchType == SpoilerCollectionCheckType::SPOILER_CHK_RANDOMIZER_INF &&
+              (scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_MERCHANT ||
+               scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_SHOP_ITEM ||
+               scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_COW ||
+               scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_SCRUB ||
+               scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_MASTER_SWORD ||
+               scCheckType == SpoilerCollectionCheckType::SPOILER_CHK_RANDOMIZER_INF)) {
+            if (flag == OTRGlobals::Instance->gRandomizer->GetRandomizerInfFromCheck(rc)) {
+                SetCheckCollected(rc);
+                return;
+            }
```
Another doesn't need to happen now, but all of these should get migrated to just using SPOILER_CHK_RANDOMIZER_INF with their associated flags, then we can skip the `GetRandomizerInfFromCheck` check.

Malkierian replied:
Yeah, this was annoying to have to deal with.

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -658,42 +598,124 @@ void CheckTrackerItemReceive(GetItemEntry giEntry) {
+            }
+            continue;
+        }
+        if (checkMatchType == FLAG_ITEM_GET_INF && flag == 0x1000) {
+            SetCheckCollected(RC_GRAVEYARD_DAMPE_GRAVEDIGGING_TOUR);
+            return;
+        }
```
Why is this one here instead of in the switch above with the others?

Malkierian replied:
Honestly have no idea XD.

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -536,10 +494,6 @@ void CheckTrackerTransition(uint32_t sceneNum) {
-    }
     doAreaScroll = true;
     previousArea = currentArea;
     currentArea = GetCheckArea();
```
This is probably where we would check for SCENE_POTION_SHOP_KAKARIKO && IS_LINK_ADULT yea?

Malkierian replied:
In reference to what? Usually, if I need to differentiate, I differentiate where the functionality differs. But thanks for reminding me that I need to prevent the shop stuff from updating adult potion shop when entering as child XD.

garrettjoecox replied:
Yea that's what this was in reference to, more so the code below my comment this was just the lowest I could place the comment

Malkierian replied:
To further answer the question, though, the RCArea will still be Kak whether you're child or adult in the Kak potion shop, so the check for age in the potion shop scene actually went in SetShopSeen.

garrettjoecox reviewed (approved):
Looks good, not blocking but a few comments.

Malkierian committed:
Fix Gravedigging Tour tracking.
Simplify Always Win Gravedigging Tour and Fix Gravedigging Tour Glitch applications.
Modified all necessary paths to use vanilla GDT PoH collection flag instead of randomizer variant.

Malkierian committed:
Fix Kak Potion Shop being "seen" when entering as child.
